crates/lib/src/execution_data.rs

```diff
@@ -63,44 +126,95 @@ impl ReadoutMap {
         readout_mappings: &HashMap<String, String>,
         readout_values: &HashMap<String, ReadoutValues>,
     ) -> Self {
-        let result = readout_values
-            .iter()
-            .flat_map(|(readout_name, values)| {
-                readout_mappings
-                    .iter()
-                    .filter(|(_, program_alias)| *program_alias == readout_name)
-                    .map(|(program_name, _)| program_name.as_ref())
-                    .map(parse_readout_register)
-                    .map(|reference| Ok((reference?, values.clone())))
-                    .collect::<Result<Vec<_>, MemoryReferenceParseError>>()
-            })
-            .flatten()
-            .collect::<HashMap<_, _>>()
-            .into();
+        let mut result = ReadoutMap(HashMap::new());
+        for (readout_name, values) in readout_values {
+            readout_mappings
+                .iter()
+                .filter(|(_, program_alias)| *program_alias == readout_name)
+                .map(|(program_name, _)| program_name.as_ref())
+                .map(parse_readout_register)
+                .filter_map(Result::ok)
+                .for_each(|reference| {
+                    let row = match &values.values {
+                        Some(readout_values::Values::IntegerValues(ints)) => ints
+                            .values
+                            .clone()
+                            .into_iter()
+                            .map(|i| Some(ReadoutValue::I32(i)))
+                            .collect(),
+                        Some(readout_values::Values::ComplexValues(comps)) => comps
+                            .values
+                            .clone()
+                            .into_iter()
+                            .map(|c| Complex64::new(c.real().into(), c.imaginary().into()))
+                            .map(|c| Some(ReadoutValue::Complex64(c)))
+                            .collect(),
+                        None => Vec::new(),
+                    };
+                    // TODO handle possible truncation
+                    let shape = (reference.index as usize + 1, row.len());
+                    let matrix = result
+                        .0
+                        .entry(reference.name)
+                        .and_modify(|m| {
+                            if shape.0 > m.nrows() {
+                                *m = Array2::from_shape_fn(shape, |(r, c)| {
+                                    *m.get((r, c)).unwrap_or(&None)
```

Comment:
What is the logic here? Looks like you're converting an `Option<Option<T>>` to an `Option<T>`. Would `Option::flatten` make more sense here?

Reply:
We visit the memory references in arbitrary order, so we initialize the matrix with dimensions using the first memory index and row length we see for a register. The logic here is to grow the matrix if we encounter a reference with a higher memory index than we've previously seen. This `from_shape_fn` should populate a new matrix with the new dimension using the values in the existing matrix, and pad in any new indices with `None`.

To answer your question, we can indeed use `flatten()` in this case, and looks cleaner than defaulting to `&None` - thanks for letting me know of its existence.